lib/miq_expression/remove_equality_visitor.rb

```diff
@@ -0,0 +1,71 @@
+class MiqExpression
+  # rubocop:disable Naming/MethodName
+  class RemoveEqualityVisitor < Arel::Visitors::Reduce
+    def visit_NilClass(object, _collector)
+      object
+    end
+
+    def visit_Arel_Nodes_Equality(object, collector)
+      field, binds = collector
+      if (object.left == field.left && object.right == field.right) ||
+         (object.right == field.left && object.left == field.right)
+        # remove this node from the query
+        nil
+      elsif object.right.kind_of?(Arel::Nodes::BindParam)
+        # replace this node with "field = $(next value from the binds array)"
+        object.left.eq(binds.shift.value_for_database)
+      else
+        # leave this node in the query unchanged
+        object
+      end
+    end
+
+    def visit_Arel_Nodes_And(object, collector)
+      children = object.children.compact.flatten.map { |q| visit(q, collector) }
+      children.compact!
+
+      if children.blank?
+        nil
+      elsif children.size == 1
+        children.first
+      elsif children == object.children
+        object
+      else
+        Arel::Nodes::And.new(children)
+      end
+    end
+
+    def visit_Arel_Nodes_Grouping(object, collector)
+      expr = visit(object.expr, collector)
+      if expr == object.expr
+        object
+      elsif expr.nil?
+        nil
+      else
+        Arel::Nodes::Grouping.new(expr)
+      end
+    end
+
+    # This method does 2 things:
+    # - remove node from the query
+    # - replace bind variables in the query
+    #
+    # Active record has the concept of the bind param values, but arel does not.
+    # So converting a query into arel will loose the values for the bind parameters.
+    # The arel becomes incomplete and can not generate valid sql.
```

Comment:
It seems like the description of this method is describing the "why", not the "how".  Specifically, I think their are some concepts that are worth mentioning:

- `collector` is just a tuple of the `node_to_remove` and the `bound_attributes` that are passed in, and not a `ActiveRecord`/`Arel` construct.  I think this is important now more than when the `5.1` code existed because it is less obvious how the `collector` is being used now.
- Both `object` and `bind` will be mutated as a result of this method, not just returned.  Correct?

Reply:
the first 2 lines tried to tell what was happening

Reply:
changed collector to an actual class. still not collecting, but at least it is a single object.
in the future we may use the visitor to add node values to the collector

Reply:
wait can i also jump on the typo train here? sorry Keenan but it should be `lose`